registry/storage/garbagecollect.go

```diff
@@ -89,25 +100,31 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 				return fmt.Errorf("failed to retrieve manifest for digest %v: %v", dgst, err)
 			}
 
-			descriptors := manifest.References()
-			for _, descriptor := range descriptors {
+			for _, descriptor := range manifest.BlobReferences() {
 				markSet[descriptor.Digest] = struct{}{}
 				emit("%s: marking blob %s", repoName, descriptor.Digest)
 			}
 
-			return nil
-		})
+			for _, descriptor := range manifest.ManifestReferences() {
+				if _, ok := unmarkedManifests[descriptor.Digest]; ok {
```

Comment:
For manifest references add it to the `manifestList` for processing if we haven't done so already.

Reply:
I'm not 100% on what you're asking here so apologies if I'm addressing the wrong concrn.

The `manifestEnumerator.Enumerate` loop at line 66 computes the initial reachable set of manifests. If RemoveUntagged is set this is only manifests that are directly tagged, otherwise it trivially just includes all manifests.

Then this loop starting at line 91 is computing the transitive closure on that initial set of manifests. Each iteration we process a reachable manifest list and mark all its blobs and manifests. If it has any manifest dependencies that have not already been marked we add it to the end of the list we're iterating over. Logically we're just executing a breadth first search.

In this way all manifests and blobs that are directly or indirectly referenced by a manifest are eventually marked.

Reply:
> we add it to the end of the list we're iterating over. Logically we're just executing a breadth first search.
>
> In this way all manifests and blobs that are directly or indirectly referenced by a manifest are eventually marked.

I see this now, it's pretty elegant. Sorry for missing it at first :+1: 

Could you add a comment above line 111 bringing attention to the fact that manifests which are referenced by manifest lists are being appended to the end of the list so that they're processed before the loop ends?